test/Fields/field.jl

```diff
@@ -468,3 +468,41 @@ end
     C .= Ref(zero(eltype(C)))
     @test all(==(0.0), parent(C))
 end
+
+# https://github.com/CliMA/ClimaCore.jl/issues/946
+@testset "Allocations with broadcasting Refs" begin
+    FT = Float64
+    function foo!(Yx::Fields.Field)
+        Yx .= Ref(1) .* Yx
+        return nothing
+    end
+    function foocolumn!(Yx::Fields.Field)
+        Fields.bycolumn(axes(Yx)) do colidx
+            Yx[colidx] .= Ref(1) .+ Yx[colidx]
+            nothing
+        end
+        return nothing
+    end
+    for space in all_spaces(FT)
+        (
+            space isa Spaces.ExtrudedFiniteDifferenceSpace ||
+            space isa Spaces.SpectralElementSpace1D ||
+            space isa Spaces.SpectralElementSpace2D
+        ) || continue
+        Y = FieldFromNamedTuple(space, (; x = FT(2)))
+
+        # Plain broadcast
+        Yx = Y.x
+        foo!(Yx) # compile first
+        p = @allocated foo!(Yx)
+        @test_broken p == 0 # 16 bytes
+        # debug info:
+        # @info "space = $(nameof(typeof(space)))"
+        # @show p
```

Comment:
cc @simonbyrne, not sure why this is still allocating, does anything stick out to you?

Reply:
Fixed 🎉